qiskit/_jobprocessor.py

```diff
@@ -0,0 +1,366 @@
+        self.seed = seed
+        self.result = None
+        self.doCompile = doCompile
+
+
+class JobProcessor():
+    """
+    process a bunch of jobs and collect the results
+    """
+    def __init__(self, q_jobs, callback, max_workers=1, token=None, url=None, api=None):
+        """
+        Args:
+            q_jobs (list(QuantumJob)): List of QuantumJob objects.
+            callback (fn(results)): The function that will be called when all
+                jobs finish. The signature of the function must be:
+                fn(results)
+                results: A list of (result, qobj) tuples.
+            max_workers (int): The maximum number of workers to use.
+            token (str): Server API token
+            url (str): Server URL.
+            api (IBMQuantumExperience): API instance to use. If set,
+                /token/ and /url/ are ignored.
+        """
+        self.q_jobs = q_jobs
+        self.max_workers = max_workers
+        # check whether any jobs are remote
+        self.online = False
+        self._local_backends = local_backends()
+        for qj in self.q_jobs:
+            if qj.backend not in self._local_backends:
+                self.online = True
+                break
+        self.futures = {}
+        self.lock = Lock()
+        # Set a default dummy callback just in case the user doesn't want
+        # to pass any callback.
+        self.callback = (lambda rs:()) if callback is None else callback
+        self.num_jobs = len(self.q_jobs)
+        self.jobs_results = []
+        if self.online:
+            self._api = api if api else IBMQuantumExperience(token,
+                                                             {"url": url},
+                                                             verify=True)
+            self._online_backends = remote_backends(self._api)
+            self._api_config = {}
+            self._api_config["token"] = token
+            self._api_config["url"] =  {"url": url}
+        else:
+            self._api = None
+            self._online_backends = None
+            self._api_config = None
+        if self.online:
+            # I/O intensive -> use ThreadedPoolExecutor
+            self.executor_class = futures.ThreadPoolExecutor
+        else:
+            # CPU intensive -> use ProcessPoolExecutor
+            self.executor_class = futures.ProcessPoolExecutor
+
+    def _job_done_callback(self, future):
+        try:
+            result = future.result()
+        except Exception as ex:
+            result = {'status': 'ERROR', 'result': '{}'.format(ex)}
+
+        with self.lock:
+            self.futures[future]['result'] = result
+            self.jobs_results.append((result, future.qobj))
+            if self.num_jobs != 0:
+                self.num_jobs -= 1
+        # Call the callback when all jobs have finished
+        if self.num_jobs == 0:
+            if not future.silent:
+                import pprint
+                pprint.pprint(f.result())
```

Comment:
Should the `f` variable be `future`?

Reply:
it's "result" indeed, but yeah, it meant to be future.